### ISSUE TITLE
Description of Build Order and bootstrapped builds

### ISSUE THREAD

**Issue by hmaarrfk** [tensorflow/tensorflow#53582]
I would like to request documentation on the ideal build order for tensorflow and both its required dependencies as well as suggested dependencies:

As described in: https://github.com/tensorflow/tensorflow/blob/v2.7.0/tensorflow/tools/pip_package/setup.py#L104
the package `tensorflow` depends on 4 dependencies that must be moved in step:
```
    'tensorboard ~= 2.6',
    'tensorflow_estimator ~= 2.7.0rc0, < 2.8',
    # Keras release is not backward compatible with old tf release, and we have
    # to make the version aligned between TF and Keras.
    'keras >= 2.7.0rc0, < 2.8',
    'tensorflow-io-gcs-filesystem >= 0.21.0',
```

However, it is my understanding that in order to build `estimator` that `tensorflow` must be importable by python.

Is that really the case?

If not, is there somewhere where we can read up on the ideal build order. Ideally we would have:
1. Build `tensorflow`. This will create an importable package, but without many extra features.
     * It might be that this package is called `tensorflow-base` or something else to your liking
2. Build `estimator`....
3. Build `keras` ....
4. Assemble all dependencies into a single dependency called `tensorflow`

Ideally this would form some directed acyclic graph so that we can bootstrap the builds without having an "older version of tensorflow" already compiled.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Any
- TensorFlow installed from (source or binary): source
- TensorFlow version: 2.7.0
- Python version: 3.7-3.10
- Installed using virtualenv? pip? conda?: conda with conda-forge
- Bazel version (if compiling from source): 4.XX
- GCC/Compiler version (if compiling from source): 9
- CUDA/cuDNN version: 11.2
- GPU model and memory: Many

**Describe the problem**

It is best to look a the file `recipe/build.sh`  that builds the wheel. Finally the file `recipe/build_pkg.sh` installs the wheel.
https://github.com/conda-forge/tensorflow-feedstock/pull/189

**Any other info / logs**
The best logs can be found on the conda-forge recipe. Here is the PR upgrading the system to 2.7.0. You can see a few patches strip out certain dependencies.

https://github.com/conda-forge/tensorflow-feedstock/pull/189

**Comment by tilakrayal** [comment]
@hmaarrfk ,
Please feel free to submit a PR for the requested change or share the link where requested change is to be made

**Comment by hmaarrfk** [comment]
Thank you for your support. I will look into how best to document things. Hopefully it won't require any changes to the source.

**Comment by tilakrayal** [comment]
@hmaarrfk ,
As suggested Please feel free to submit a PR for the requested change.Thanks!

**Comment by hmaarrfk** [comment]
I'm not sure why this is stale. Can we mark it as not stale? This kind of documentation is not a trivial one liner.

In my mind, the solution is one of two things:
1. Documentation including "patches" to apply to `tensorflow` and dependencies such as `tensorflow-io` to get the system to compile from a boostrapped environment.
2. Creating new packages that break the circular dependency between `tensorflow` and `tensorflow-io`.

For example, see how tensorflow depends on `tensorflow-io-gcs-filesystem`:
https://github.com/tensorflow/tensorflow/blob/v2.7.0/tensorflow/tools/pip_package/setup.py#L109

And how `tensorflow-io-gcs-filesystem` depends on tensorflow:
https://github.com/tensorflow/io/blob/v0.23.1/setup.py#L171
https://github.com/tensorflow/io/blob/v0.23.1/tensorflow_io/python/ops/version_ops.py#L18

Typically, these circular dependencies are broken by having a "core" package. The DAG might look like this
```
tensorflow ---------------------------------------------> tensorflow-core
          \                                              ^
           \                                            /
            \--------tensorflow-io-gcs-filesystem -----/
```
Having such a DAG allows people to have a clear starting point of where compilation should start, and what packages need to be built sequentially.

At conda-forge, we are getting close to doing option `1.` but, it is rather unmaintainable (and not yet fully documented)

Option 2, in my mind, is in the interest of `tensorflow` developers as untangling the dependencies will help them maintain more architectures.

Option #2 is more than what I can provide. and why I think it is an important issue to raise for tensorflow at large.

**Comment by mihaimaruseac** [comment]
Our `BUILD` files ensure that there is no actual cyclic dependency between the projects, except at Python import layer, as modules/plugins.

Hence, you can build TF, Estimator (deprecated, will be dropped in the future), the GCS filesystem plugin in this order and then at import time in TF it all should work.

We do this when building the standard release:

- cut branch in TF, cut branch in Keras and Estimator at commits that corresponds to the same point in time (`PiperOrigin-RevId` number is close to the one in the TF's commit but not greater)
- build TF RC0 against estimator nightly wheel (we can build it without any estimator dependency but then we won't be able to run all tests or we would need a separate script for this)
- built Estimator RC0 against TF RC0
- Keras RC0 can be build in parallel, it does not depend on Estimator and C++ parts of TF
- update bounds in TF's `setup.py` to use these RC0s
- do TF RC1, RC2, etc and similar for the others if needed, against latest TF RC
- release final Estimator against latest TF RC
- release final Keras
- update bounds in TF's `setup.py` to use the latest versions
- release final TF

We had a proposal 2 years ago to modularize all this and have a real DAG of dependencies but there has been significant pushback as that would break the ability to do cross-cutting changes that some C++ teams preferred to do. Then, people that proposed the modularization left the team, the team halved, the proposal dropped.

**Comment by hmaarrfk** [comment]
Thank you for explination of the build process. I think this should help us better put things together.

In your description, does "GCS filesystem" get built in parallel with estimator?

**Comment by mihaimaruseac** [comment]
The GCS filesystem is built by SIG IO community. It can be build in parallel with any other TF process, it doesn't need TF afaik.

The filesystem is a shared object (`.so`, `.dylib`, `.dll`, depending on operating system) that is then loaded by TF on import,if present.

**Comment by hmaarrfk** [comment]
Not stale.

**Comment by SuryanarayanaY** [comment]
Hi @hmaarrfk ,

With latest version tf-estimator was deprecated and no longer needed for build and refer the [source](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/pip_package/setup.py#L84).

GCS filesystem is built by SIG IO community as communicated in [comment-1031094136] (https://github.com/tensorflow/tensorflow/issues/53582#issuecomment-1031094136).

Could you please confirm whether you still working on this issue ? Thanks!

**Comment by h-vetinari** [comment]
Hi @SuryanarayanaY, yes this is still important in conda-forge. Can you

> With latest version tf-estimator was deprecated and no longer needed for build and refer the [source](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/pip_package/setup.py#L84).

I'm not sure I understand. It's still listed under `REQUIRED_PACKAGES` for [2.12](https://github.com/tensorflow/tensorflow/blob/v2.12.0-rc1/tensorflow/tools/pip_package/setup.py#L130-L131) and on the master branch.

**Comment by mihaimaruseac** [comment]
Estimator does not receive more updates, so will likely be deprecated. However, it is likely that for a period of time one last build of Estimator would still be needed for TF.

**Comment by hmaarrfk** [comment]
Not stale.

**Comment by SuryanarayanaY** [comment]
@hmaarrfk ,

Are you proposing any PR for your thoughts?

Right now to build Tensorflow from source, I used to do pip install the required packages mentioned under setup.py following the instructions mentioned [here](https://www.tensorflow.org/install/source).

**Comment by hmaarrfk** [comment]
I'm not sure why you are asking me for a PR when the discussion hints that bootstrapping is just really not that easy as explained in the comment https://github.com/tensorflow/tensorflow/issues/53582#issuecomment-1030208849 

I am really not the best person for this since I am merely a user of tensorflow. Rather, i think it is more appropriate for somebody intimately knowledgeable with tensorflow to draw a dependency graph. In doing so, hopefully the person will realize how it is nearly impossible to build tensorflow if you have to bootstrap the build process and maybe this may motivate some dependency cleanup 

However, that isn't easy, and it seems the priority isn't there.

I'm not pinging your team asking for this to be resolved, i understand i don't drive your development plan. But i do think that this issue is still relevant. However, when asked if it is still relevant, i will kindly say that it still is pleading for the issue to remain open.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by mihaimaruseac** [comment]
There is a push to close all issues, there is a metric on old issues that needs to be pushed down. But the implementation is lacking, some issues are documentation or long process, whereas others are simple bugs. Right now, they're all bundled together and only getting the metric down matters.

**Comment by h-vetinari** [comment]
> only getting the metric down matters.

This only ends up papering over problems and alienating users/contributiors who get their issues closed without engaging with (much less fixing) the underlying problem, but look KPIs!

(Not directed at either of you directly, but it's just such a shortsighted way to manage a project IMO 😑).
